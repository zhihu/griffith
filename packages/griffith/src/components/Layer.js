import React from 'react'
import {css} from 'aphrodite/no-important'
import PositionContext from '../contexts/PositionContext'
import ObjectFitContext from '../contexts/ObjectFitContext'
import styles from './Layer.styles'

const getContainerClassName = isFullWidth =>
  css(
    styles.container,
    isFullWidth ? styles.containerFullWidth : styles.containerFullHeight
  )

const getImageClassName = isFullWidth =>
  css(
    styles.image,
    isFullWidth ? styles.imageFullWidth : styles.imageFullHeight
  )

const Positioned = ({children}) => (
  <PositionContext.Consumer>
    {({isFullWidth, helperImageSrc}) =>
      helperImageSrc && (
        <div className={getContainerClassName(isFullWidth)}>
          <img
            src={helperImageSrc}
            className={getImageClassName(isFullWidth)}
          />
          {children}
        </div>
      )
    }
  </PositionContext.Consumer>
)

export default function Layer({children}) {
  if (!children) return null

  const layer = <div className={css(styles.layer)}>{children}</div>

  // 暂时先只考虑 cover
  return (
    <ObjectFitContext.Consumer>
      {({objectFit}) =>
        objectFit === 'cover' ? layer : <Positioned>{layer}</Positioned>
      }
    </ObjectFitContext.Consumer>
  )
}
