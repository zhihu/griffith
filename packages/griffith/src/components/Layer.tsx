import React, {useContext} from 'react'
import {css} from 'aphrodite/no-important'
import PositionContext from '../contexts/PositionContext'
import ObjectFitContext from '../contexts/ObjectFitContext'
import styles from './Layer.styles'

const getContainerClassName = (isFullWidth: boolean) =>
  css(
    styles.container,
    isFullWidth ? styles.containerFullWidth : styles.containerFullHeight
  )

const getImageClassName = (isFullWidth: boolean) =>
  css(
    styles.image,
    isFullWidth ? styles.imageFullWidth : styles.imageFullHeight
  )

const Positioned: React.FC = ({children}) => {
  const {isFullWidth, helperImageSrc} = useContext(PositionContext)

  if (helperImageSrc) {
    return (
      <div className={getContainerClassName(isFullWidth)}>
        <img src={helperImageSrc} className={getImageClassName(isFullWidth)} />
        {children}
      </div>
    )
  }

  return null
}

const Layer: React.FC = ({children}) => {
  const {objectFit} = useContext(ObjectFitContext)

  if (children) {
    const layer = <div className={css(styles.layer)}>{children}</div>
    // 暂时先只考虑 cover
    return objectFit === 'cover' ? layer : <Positioned>{layer}</Positioned>
  }

  return null
}

export default Layer
