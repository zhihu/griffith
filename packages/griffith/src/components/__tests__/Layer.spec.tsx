import React from 'react'
import {shallow} from 'enzyme'
import Layer from '../Layer'

describe('Layer', () => {
  it('get Layer component', () => {
    const wrapper1 = shallow(<Layer />)
    expect(wrapper1).toEqual({})

    const wrapper2 = shallow(
      <Layer>
        <span>123</span>
      </Layer>
    )
    expect(wrapper2).toMatchSnapshot()
  })
})
