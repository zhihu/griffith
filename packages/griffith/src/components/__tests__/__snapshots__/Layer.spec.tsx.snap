// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layer get Layer component 1`] = `
<Positioned>
  <div
    className="layer_120lr3k"
  >
    <span>
      123
    </span>
  </div>
</Positioned>
`;
